src/argilla/client/feedback/training/schemas.py

```diff
@@ -154,49 +159,74 @@ def for_text_classification(
+        elif text or label:
+            raise ValueError("You must define both `text` and `label`.")
+        elif not (text or label) and formatting_func is None:
+            raise ValueError("You must define either (`text` and `label`), or a `formatting_func`.")
+
+        if formatting_func is not None:
+            if text or label:
+                raise ValueError("`formatting_func` is already defined, so you cannot define `text` and `label`.")
+            return TrainingTaskForTextClassification(formatting_func=formatting_func)
         else:
-            unification_kwargs = {"question": label}
-            if label_strategy is not None:
-                unification_kwargs["strategy"] = label_strategy
-            else:
-                _LOGGER.info(f"No label strategy defined. Using default strategy for {type(label)}.")
-            if isinstance(label, RatingQuestion):
-                label = RatingQuestionUnification(**unification_kwargs)
-            elif isinstance(label, MultiLabelQuestion):
-                label = MultiLabelQuestionUnification(**unification_kwargs)
-            elif isinstance(label, LabelQuestion):
-                label = LabelQuestionUnification(**unification_kwargs)
-            elif isinstance(label, RankingQuestion):
-                label = RankingQuestionUnification(**unification_kwargs)
+            if isinstance(
+                label,
+                (
+                    LabelQuestionUnification,
+                    MultiLabelQuestionUnification,
+                    RatingQuestionUnification,
+                    RankingQuestionUnification,
+                ),
+            ):
+                if label_strategy is not None:
+                    raise ValueError("label_strategy is already defined via Unification class.")
             else:
-                raise ValueError(f"Label type {type(label)} is not supported.")
-        return TrainingTaskForTextClassification(
-            text=text,
-            label=label,
-            label_strategy=label_strategy,
-        )
+                unification_kwargs = {"question": label}
+                if label_strategy is not None:
+                    unification_kwargs["strategy"] = label_strategy
+                else:
+                    _LOGGER.info(f"No label strategy defined. Using default strategy for {type(label)}.")
+                if isinstance(label, RatingQuestion):
+                    label = RatingQuestionUnification(**unification_kwargs)
+                elif isinstance(label, MultiLabelQuestion):
+                    label = MultiLabelQuestionUnification(**unification_kwargs)
+                elif isinstance(label, LabelQuestion):
+                    label = LabelQuestionUnification(**unification_kwargs)
+                elif isinstance(label, RankingQuestion):
+                    label = RankingQuestionUnification(**unification_kwargs)
```

Comment:
If there's any field and/or value that can be used to distinguish those, we could use a `Discriminator` based on that value to avoid having these if-else blocks. More information at https://docs.pydantic.dev/1.10/usage/types/#nested-discriminated-unions

Reply:
This can only be done based on literals but given that the discriminator is a Question, it wouldn't work.